include/deal.II/lac/petsc_block_vector.h

```diff
@@ -454,7 +511,12 @@ namespace PETScWrappers
     inline const MPI_Comm &
     BlockVector::get_mpi_communicator() const
     {
-      return block(0).get_mpi_communicator();
+      static MPI_Comm comm = PETSC_COMM_SELF;
+      MPI_Comm        pcomm =
+        PetscObjectComm(reinterpret_cast<PetscObject>(petsc_nest_vector));
+      if (pcomm != MPI_COMM_NULL)
+        comm = pcomm;
+      return comm;
```

Comment:
I don't think this works as expected with the `static`. Why not do this:
```suggestion
      MPI_Comm        pcomm =
        PetscObjectComm(reinterpret_cast<PetscObject>(petsc_nest_vector));
      if (pcomm != MPI_COMM_NULL)
        return pcomm;
      else
        return PETSC_COMM_SELF;
```
But one can also ask whether it is not a failure case if the call to `PetscObjectComm` returns `MPI_COMM_NULL`, and just error out in that case with an assertion.

Reply:
It needs `static` because the return argument is a reference to an `MPI_Comm` (I haven't changed it). Your code will compile with warnings, while the one with `static` is ok. Anyway, we will remove the reference from the return argument in a follow-up PR.

Reply:
regarding the `MPI_COMM_NULL` comment: If the PETSc object is `NULL`, then `PetscObjectComm` returns `MPI_COMM_NULL`. It seemed to me that you use SELF when an object is not yet set up to be used in deal.II, that's why it is returning self. That was the rationale, I can change the behavior if you prefer.

Reply:
This change will be part of the follow-up PRs, so I guess we can consider this discussion resolved.

Reply:
@bangerth Here you have the relevant commit https://github.com/dealii/dealii/pull/14481/commits/0db14d853d927f7dbb45918d71ebbc0c33dd10a9